backbone/src/main/java/org/researchstack/backbone/ui/ConsentTaskActivity.kt

```diff
@@ -0,0 +1,219 @@
+package org.researchstack.backbone.ui
+
+import android.app.Activity
+import android.app.AlertDialog
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import org.koin.androidx.viewmodel.ext.android.viewModel
+import org.koin.core.parameter.parametersOf
+import org.researchstack.backbone.R
+import org.researchstack.backbone.answerformat.BirthDateAnswerFormat
+import org.researchstack.backbone.answerformat.TextAnswerFormat
+import org.researchstack.backbone.result.StepResult
+import org.researchstack.backbone.step.ConsentDocumentStep
+import org.researchstack.backbone.step.ConsentSignatureStep
+import org.researchstack.backbone.step.FormStep
+import org.researchstack.backbone.step.QuestionStep
+import org.researchstack.backbone.step.Step
+import org.researchstack.backbone.task.Task
+import androidx.lifecycle.Observer
+import org.researchstack.backbone.ui.step.layout.ConsentSignatureStepLayout.KEY_SIGNATURE
+import org.researchstack.backbone.ui.task.TaskActivity
+import org.researchstack.backbone.ui.task.TaskViewModel
+import org.researchstack.backbone.utils.LocaleUtils
+import org.researchstack.backbone.utils.RSHTMLPDFWriter
+import java.text.DateFormat
+import java.util.*
+import kotlin.collections.ArrayList
+
+class ConsentTaskActivity : TaskActivity() {
+    private val viewModel: TaskViewModel by viewModel { parametersOf(intent) }
+    private var consentHtml: String? = null
+    private var firstName: String? = null
+    private var lastName: String? = null
+    private var signatureBase64: String? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel.saveStepEvent.observe(this, Observer {
+            onSaveStep(it.first, it.second)
+        })
+    }
+
+    private fun onSaveStep(step: Step, result: StepResult<*>?) {
+        if (step is FormStep && step.getIdentifier().equals(ID_FORM, true)) {
+            for (question in step.getFormSteps()!!) {
+                if (question.identifier.equals(ID_FORM_FIRST_NAME, true)) {
+                    val nameResult = result?.getResultForIdentifier(ID_FORM_FIRST_NAME) as StepResult<*>?
+                    if (nameResult != null && nameResult.result != null) {
+                        firstName = nameResult.result as String
+                    }
+                }
+                if (question.identifier.equals(ID_FORM_LAST_NAME, true)) {
+                    val nameResult = result?.getResultForIdentifier(ID_FORM_LAST_NAME) as StepResult<*>?
+                    if (nameResult != null && nameResult.result != null) {
+                        lastName = nameResult.result as String
+                    }
+                }
+            }
+        }
+        if (step is ConsentSignatureStep) {
```

Comment:
Is this check needed here? If the step is FormStep, then it cannot be ConsentSignature or ConsentDocumentSteps correct? Shouldn't this be part of an `else` (or `else if`) from the above if?

Perhaps this method can be cleared and made simpler to understand with less nesting. (Low priority)